api/kv.go

```diff
@@ -0,0 +1,328 @@
+	CustomMetadata map[string]string `mapstructure:"custom_metadata"`
+}
+
+func (c *Client) KVv1(mountPath string) *KVClient {
+	return &KVClient{c: c, mountPath: mountPath, version: 1}
+}
+
+func (c *Client) KVv2(mountPath string) *KVClient {
+	return &KVClient{c: c, mountPath: mountPath, version: 2}
+}
+
+func (kv *KVClient) Read(ctx context.Context, secretPath string) (*KVSecret, error) {
+	pathToRead, err := kv.getFullPath(secretPath)
+	if err != nil {
+		return nil, fmt.Errorf("error assembling full path to KV secret: %v", err)
+	}
+
+	secret, err := kv.c.Logical().ReadWithContext(ctx, pathToRead)
+	if err != nil {
+		return nil, fmt.Errorf("error encountered while reading secret at %s: %v", pathToRead, err)
+	}
+	if secret == nil || secret.Data == nil {
+		return nil, fmt.Errorf("no secret found at %s", pathToRead)
+	}
+
+	return extractDataAndVersionMetadata(secret, kv.version)
+}
+
+// ReadVersion returns the data and metadata for a specific version of the
+// given secret. If that version has been deleted, the Data field on the
+// returned secret will be nil, and the Metadata field will contain the deletion time.
+func (kv *KVClient) ReadVersion(ctx context.Context, secretPath string, version int) (*KVSecret, error) {
+	pathToRead, err := kv.getFullPath(secretPath)
+	if err != nil {
+		return nil, fmt.Errorf("error assembling full path to KV secret: %v", err)
+	}
+
+	queryParams := map[string][]string{"version": {strconv.Itoa(version)}}
+	secret, err := kv.c.Logical().ReadWithDataWithContext(ctx, pathToRead, queryParams)
+	if err != nil {
+		return nil, err
+	}
+	if secret == nil || secret.Data == nil {
+		return nil, fmt.Errorf("no secret version found at %s", pathToRead)
+	}
+
+	return extractDataAndVersionMetadata(secret, kv.version)
+}
+
+func (kv *KVClient) ReadMetadata(ctx context.Context, secretPath string) (*KVMetadata, error) {
+	pathToRead, err := kv.getFullMetadataPath(secretPath)
+	if err != nil {
+		return nil, fmt.Errorf("error assembling full path to KV secret's metadata: %v", err)
+	}
+
+	secret, err := kv.c.Logical().ReadWithContext(ctx, pathToRead)
+	if err != nil {
+		return nil, err
+	}
+	if secret == nil || secret.Data == nil {
+		return nil, fmt.Errorf("no secret metadata found at %s", pathToRead)
+	}
+
+	return extractFullMetadata(secret)
+}
+
+func (kv *KVClient) Write(ctx context.Context, secretPath string, data map[string]interface{}) (*KVSecret, error) {
+	pathToWriteTo, err := kv.getFullPath(secretPath)
+	if err != nil {
+		return nil, fmt.Errorf("error assembling full path to KV secret: %v", err)
+	}
+
+	secret, err := kv.c.Logical().WriteWithContext(ctx, pathToWriteTo, data)
+	if err != nil {
+		return nil, fmt.Errorf("error writing secret to %s: %v", pathToWriteTo, err)
+	}
+	if kv.version == 1 {
+		return nil, nil // v1 Logical Write returns a nil secret
+	}
+	if secret == nil || secret.Data == nil {
+		return nil, fmt.Errorf("no secret was written to %s", pathToWriteTo)
```

Comment:
What would be the cause of this error?

Reply:
I don't know that there's anything in the v2 backend that would cause it today (at least according to pathDataWrite [here](https://github.com/hashicorp/vault-plugin-secrets-kv/blob/master/path_data.go) which I believe is the write-handling code), but now I just have this check there for future-proofing to avoid a nil pointer deref when I call secret.Data below.